text/0000-stabilize-marker-freeze.md

```diff
@@ -0,0 +1,149 @@
+in read-only static memory or writeable static memory, or the optimizations that may be performed
+in code that holds an immutable reference to `T`.
+```
+
+Mention could be added to `UnsafeCell` and atomics that adding one to a previously `Freeze` type without an indirection (such as a `Box`) is a SemVer hazard, as it will revoke its implementation of `Freeze`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Some people have previously argued that this would be akin to exposing compiler internals.
+	- The RFC author disagrees, viewing `Freeze` in a similar light as `Send` and `Sync`: a trait that allows soundness requirements to be proven at compile time.
+- `Freeze` being an auto-trait, it is, like `Send` and `Sync` a sneaky SemVer hazard.
+	- Note that this SemVer hazard already exists through the existence of static-promotion, as exemplified by the following example:
+	```rust
+    // old version of the crate.
+    mod v1 {
+        pub struct S(i32);
+        impl S {
+            pub const fn new() -> Self { S(42) }
+        }
+    }
+
+    // new version of the crate, adding interior mutability.
+    mod v2 {
+        use std::cell::Cell;
+        pub struct S(Cell<i32>);
+        impl S {
+            pub const fn new() -> Self { S(Cell::new(42)) }
+        }
+    }
+
+    // Old version: builds
+    const C1: &v1::S = &v1::S::new();
+    // New version: does not build
+    const C2: &v2::S = &v2::S::new();
+	```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The benefits of stabilizing `core::mem::Freeze` have been highlighted in [Motivation](#motivation).
+	- By not stabilizing `core::mem::Freeze` in trait bounds, we are preventing useful and sound code patterns from existing which were previously supported.
+- Alternatively, a non-auto sub-trait of `core::mem::Freeze` may be defined:
+	- While this reduces the SemVer hazard by making its breakage more obvious, this does lose part of the usefulness that `core::mem::Freeze` would provide to projects such as `zerocopy`.
+	- A "perfect" derive macro should then be introduced to ease the implementation of this trait. A lint may be introduced in `clippy` to inform users of the existence and applicability of this new trait.
+
+# Prior art
+[prior-art]: #prior-art
+
+- This feature has been available in `nightly` for 7 years, and is used internally by the compiler.
+- The work necessary for this RFC has already been done and merged in [this PR](https://github.com/rust-lang/rust/issues/121675), and a [tracking issue](https://github.com/rust-lang/rust/issues/121675) was opened.
+- zerocopy's [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.11/zerocopy/trait.Immutable.html) seeks to provide the same guarantees as `core::marker::Freeze`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+[Should the trait be exposed under a different name?](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148)
```

Comment:
Something that seems troublesome about `Freeze` is that it's a trait named using a verb, which is good in isolation, but that verb would usually be an _operation you can perform_ on values of the implementing type. What's the “freeze” operation on `T: Freeze`? I suppose it's taking a `&` reference to `T`. That's reasonable enough, _but_ there is another “freeze” operation that's both already named in LLVM and [recently been proposed for addition to Rust](https://github.com/rust-lang/rfcs/pull/3605): “freezing” uninitialized memory to enable safely reading the bytes of a value that contains padding, or reading a value that may be partially initialized. That's a *different thing in the same problem domain* (what you can or can't do with the bytes when you have a `T` or `&T`) which seems dangerously confusing. And outside of that name collision, “`Freeze`” doesn't convey much to the naïve reader.

Unfortunately, I don't have any _good_ ideas for alternatives, especially not [non-negated ones as @RalfJung suggested](https://github.com/rust-lang/rust/pull/121501#issuecomment-1962900148), but I think that stabilizing this under the name `Freeze` would create a frequent source of confusion among those encountering it for the first time. It might be better to have an awkward complex name than a cryptic simple one. Awkward suggestion: `ShallowImmut`?

Reply:
what about `Stuck`? that is unused afaict so people seeing it would think to look up what it means in Rust. It's also nice and short and reminds users of what it means.

Reply:
At least on the theory side, a program being "stuck" has meaning already -- it means that execution can't progress further as the program has left the realm of well-defined executions.

Reply:
> At least on the theory side, a program being "stuck" has meaning already -- it means that execution can't progress further as the program has left the realm of well-defined executions. I would find this term very confusing.

I was thinking more of that the value in memory is stuck at whatever bytes are put into `.rodata`. because `Stuck` is a trait and applies to types, not program execution, it seems different enough to me that it won't be that confusing.

> Maybe we shouldn't call that "Freeze"? It's not a great name as what the operation really does is turn poison into non-deterministic regular bytes. Rust doesn't have undef-style "bytes that are unstable".

since the LLVM IR op is already called freeze, they've effectively already laid claim to that name, so anyone who's used LLVM IR and tries to figure out rust will be confused if we use `freeze` to mean something completely different where it could be confused with an operation (since freeze is present-tense, unlike stuck where being past-tense and being applied to types makes it much less likely to be an operation).

Reply:
> I was thinking more of that the value in memory is stuck at whatever bytes are put into .rodata

That's far from it, though. `Freeze` isnt just about `rodata`, it's about all shared references without interior mutability.

I find "stuck" extremely unintuitive. It conveys negative feelings to me, "we're stuck somewhere or with something (and want to get unstuck again)". I don't think it is a good term for "this memory is immutable".

> since the LLVM IR op is already called freeze, they've effectively already laid claim to that name, so anyone who's used LLVM IR and tries to figure out rust will be confused if we use freeze to mean something completely different where it could be confused with an operation (since freeze is present-tense, unlike stuck where being past-tense and being applied to types makes it much less likely to be an operation).

LLVM also uses poison/undef, neither of which we use in Rust (we call it "uninitialized memory" or just "uninit"). LLVM calls it `getelementptr` or `ptraddr`, we call it `offset`. LLVM calls it load/store, we call it read/write. The list can probably go on for a while. I don't think we should let LLVM dictate our terminology.

Reply:
> LLVM also uses poison/undef, neither of which we use in Rust (we call it "uninitialized memory" or just "uninit"). LLVM calls it `getelementptr` or `ptraddr`, we call it `offset`. LLVM calls it load/store, we call it read/write. The list can probably go on for a while. I don't think we should let LLVM dictate our terminology.

yes, I have no problem using a different name than LLVM, what I have a problem with is using the same name as LLVM but for a different operation. (yes, I think naming `f32::muladd` was a mistake).

Reply:
In zerocopy, we went with [`Immutable`](https://docs.rs/zerocopy/0.8.0-alpha.12/zerocopy/trait.Immutable.html), which I quite like (credit to @jswrenn's coworker for suggesting the name). It is an adjective and accurately describes the semantics - `T: Immutable` means that `T` cannot be mutated behind a shared reference.

Reply:
but `T` can be mutated behind a shared reference if that mutation is hidden in some other block of memory: `Box<Cell<u8>>` does implement `Freeze`, but can be mutated behind a shared reference. therefore I think using `Immutable` would be highly misleading.

Reply:
Arguably *any* name that doesn't clarify that the property only applies shallowly/not-via-indirection is similarly misleading. If we're worried about that, we'd need to add an adjective like `Shallow` to the name.

Reply:
Maybe just invent another term that no one else used. `Freeze` itself is an invented term anyway. 

Perhaps `Solid` (you "freeze" liquid to get "solid" :shrug:). 
Or `Acellular` (technically a negated word).

But given that the trait has existed for so long under the name `Freeze` perhaps it has already gotten traction 

Though IMO it's fine to actually use a multi-word term like `ShallowImmutable` or `CellFree`? We already have got the the two-word `UnwindSafe` and three-word `RefUnwindSafe` auto traits in std. As long as this trait bound is not needed very often (unlike `Send` and `Sync` and `Sized`) using a longer but more precise name shouldn't be a problem.

Reply:
Let me set a record for the most negative name: `UnindirectedUnUnsafeCelled` :smile: 

I like `ShallowImmutable` a lot, as it's rather descriptive and attracts curiosity thanks to the `Shallow` term.

Do we have any opposition to `ShallowImmutable`?